### ISSUE TITLE
Weirdness with subclass constructors in es6.classes transformer

### ISSUE THREAD

**Issue by Alhadis** [babel/babel#2552]
This is difficult to explain, so I've written an isolated test case:

```js
let instances	= new Map();

class Resource{

	static get(id){
		let list = instances.get(this) || {};
		return list[id];
	}

	static add(id, instance){
		let list = instances.get(this);
		if(!list)
			instances.set(this, list = {});
		list[id] = instance;
	}

	constructor(id){
		let type		= this.constructor;
		let byId		= type.get(id);

		// Already created. Return reference to existing instance.
		if(byId){
			return byId;
		}

		this.id = id;
		type.add(id, this);
	}
}


class SubtypeResource extends Resource{
	
}


let resourceA = new SubtypeResource(200);
let resourceB = new SubtypeResource(200);
console.log(resourceA === resourceB); // "true" if natively supported; "false" if transpiled
```

Blacklisting `es6.classes` solves the problem, but that naturally requires an interpreter to support it natively. I'm running the latest version of Node (4.2.1), so it's no big deal. But it'd be different if it were for client-side delivery.

This only happens with subclasses, so it might have something to do with the way the transformer maps references to constructor functions. E.g., the following code runs fine either natively or through Babel:
```js
console.log(new Resource(200) === new Resource(200)); // true
``` 

----
To summarise what my code's doing: I'm writing a Node program to scrape data from a site that lacks a formal API. Resources are loaded once and done so asynchronously, collating patches of data from numerous different sources. I thought an elegant approach would be to return references to existing instances instead of manually checking if they've been created already:

```js
let instance;
if(SubtypeResource.get(id)){
	instance = SubtypeResource.get(id);
}
else{
	instance = new SubtypeResource(id);
	SubtypeResource.add(id, instance);
}
```

**Comment by zloirock** [comment]
Problem here in parent  constructor calling.
```js
    function SubtypeResource() {
        _classCallCheck(this, SubtypeResource);

        _get(Object.getPrototypeOf(SubtypeResource.prototype), "constructor", this).apply(this, arguments);
    }
```
`babel` just applied parent constructor on instance and ignored result, but should use as `this` result if it's object and, in this case, return it from constructor by default. Something like
```js
    function SubtypeResource() {
        _classCallCheck(this, SubtypeResource);
        var _result = _get(Object.getPrototypeOf(SubtypeResource.prototype), "constructor", this).apply(this, arguments);
        var _this = _isObject(_result) ? _result : this;
        return _this;
    }
```
It can cause many problems and related [this](https://github.com/babel/babel/issues/1172) old conception.

**Comment by sebmck** [comment]
Duplicate of #2279.
